bastion.cfn.yml

```diff
@@ -16,68 +16,290 @@ Parameters:
+                if [ "${MFA}" == "true" ]; then
+                  echo "auth       required     pam_google_authenticator.so nullok" >> /etc/pam.d/sshd
+                  sed -e '/auth       substack     password-auth/ s/^#*/#/' -i /etc/pam.d/sshd
+                  sed -e '/ChallengeResponseAuthentication no/ s/^#*/#/' -i /etc/ssh/sshd_config
+                  sed -e '/#ChallengeResponseAuthentication yes/s/^#//' -i /etc/ssh/sshd_config
+                  echo >> /etc/ssh/sshd_config
+                  echo "AuthenticationMethods publickey,keyboard-interactive" >> /etc/ssh/sshd_config
+                  service sshd restart
+                fi
+                rm -f /usr/local/sbin/configure_mfa.sh
+
+          commands:
+            configure-mfa:
+              command: "/usr/local/sbin/configure_mfa.sh"
+
+          services:
+              sysvinit:
+                cfn-hup:
+                  enabled: true
+                  ensureRunning: true
+                  files:
+                    - /etc/cfn/cfn-hup.conf
+                    - /etc/cfn/hooks.d/cfn-auto-reloader.conf
+                awslogs:
+                  enabled: true
+                  ensureRunning: true
+                  files:
+                    - /etc/awslogs/awslogs.conf
+
     Properties:
       InstanceType: t2.micro
       KeyName: !Ref KeyName
+      NetworkInterfaces:
+        -
+          NetworkInterfaceId: !Ref BastionNetworkInterface
+          DeviceIndex: 0
+
       ImageId: !FindInMap [AMIMap, !Ref "AWS::Region", AMI]
-      SubnetId: !ImportValue
-        "Fn::Sub": "${NetworkStackName}-PublicSubnet1ID"
-      SecurityGroupIds:
-      - !ImportValue
-        "Fn::Sub": "${NetworkStackName}-BastionGroupID"
+      UserData:
+        Fn::Base64: !Sub |
+          #!/bin/bash -xe
+          yum update -y
+          /opt/aws/bin/cfn-init -v -s ${AWS::StackId} --resource BastionHost --region ${AWS::Region}
+          /opt/aws/bin/cfn-signal -e $? --stack ${AWS::StackId} --resource BastionHost --region ${AWS::Region}
+
+      IamInstanceProfile: !Ref BastionInstanceProfile
       Tags:
         -
           Key: Name
           Value: startup-kit-bastion
+    DependsOn: BastionSshPort
+    CreationPolicy:
+      ResourceSignal:
+        Count: 1
+        Timeout: PT5M
 
   BastionEIP:
     Type: AWS::EC2::EIP
     Properties:
-      InstanceId: !Ref BastionHost
       Domain: vpc
 
+  BastionSshPort:
+    Type: AWS::EC2::EIPAssociation
+    Properties:
+      AllocationId: !GetAtt BastionEIP.AllocationId
+      NetworkInterfaceId: !Ref BastionNetworkInterface
+    DependsOn:
+      - BastionEIP
+      - BastionNetworkInterface
+
+  BastionNetworkInterface:
+    Type: AWS::EC2::NetworkInterface
+    Properties:
+      SubnetId:
+        Fn::ImportValue: !Sub "${NetworkStackName}-PublicSubnet1ID"
+      GroupSet:
+        - Fn::ImportValue: !Sub "${NetworkStackName}-BastionGroupID"
+      SourceDestCheck: true
+      Tags:
+        -
```

Comment:
Nit: For consistency, settle on whether to have a line break at the start of a collection. See vs. line 263.

Reply:
Single item versus object arrays. I could go either way on this.

Reply:
No. These are collections of objects which represent tags. The consistency nit I had was the line break after the hyphen which isn't uniformly applied throughout the template.